draft-ietf-quic-multipath.md

```diff
@@ -1232,6 +1264,63 @@ a PATH_RESPONSE frame in order to indicate the preferred path usage
 before or during path initiation.
 
 
+## MP_NEW_CONNECTION_ID frames {#mp-new-conn-id-frame}
+
+An endpoint sends a MP_NEW_CONNECTION_ID frame (type=0x15228c09) instead of 
+the NEW_CONNECTION_ID frame to provide its peer with alternative connection IDs 
+that can be used to break linkability when migrating connections; 
+see {{Section 19.15 of QUIC-TRANSPORT}}.
+
+MP_NEW_CONNECTION_ID frames are formatted as shown in {{fig-mp-connection-id-frame-format}}.
+
+~~~
+MP_NEW_CONNECTION_ID Frame {
+  Type (i) = 0x15228c09,
+  Sequence Number (i),
+  Retire Prior To (i),
+  Path Identifier (i),
+  Length (8),
+  Connection ID (8..160),
+  Stateless Reset Token (128),
+}
+~~~
+{: #fig-mp-connection-id-frame-format title="MP_NEW_CONNECTION_ID Frame Format"}
+
+MP_NEW_CONNECTION_ID frames contain the following fields:
+
+Path Identifier:
+: A path identifier which is pre allocated when the Connection ID is generated, which
+means the current Connection ID can only be used on the corresponding path.
+
+
+
+## MP_RETIRE_CONNECTION_ID frames {#mp-retire-conn-id-frame}
+
+An endpoint sends a MP_RETIRE_CONNECTION_ID frame (type=0x15228c0a) instead of 
+RETIRE_CONNECTION_ID frame to indicate that it will no longer use a connection ID 
+that was issued by its peer. This includes the connection ID provided during the handshake. 
+Sending a MP_RETIRE_CONNECTION_ID frame also serves as a request to the peer 
+to send additional connection IDs for future use. New connection IDs can be 
+delivered to a peer using the MP_NEW_CONNECTION_ID frame ({{mp-new-conn-id-frame}}).
+
+Retiring a connection ID invalidates the stateless reset token associated with that connection ID.
+
+MP_RETIRE_CONNECTION_ID frames are formatted as shown in {{fig-mp-retire-connection-id-frame-format}}.
+
+~~~
+RETIRE_CONNECTION_ID Frame {
+  Type (i) = 0x15228c0a,
+  Sequence Number (i),
+  Path Identifier (i),
```

Comment:
Same here.

Reply:
Actually I think we should keep the Sequence Number of CID unique for the whole connection, although the CID is allocated for each path. Endpoints use the sequence number to find the corresponding CID.  
As when I went through the implementation, I found endpoints need to maintain all the unused CIDs under a connection CID set before it can create any new path. 

That's a important point that we shouldn't change from RFC 9000, and that's why I prefer to keep the Path Identifier field after the Sequence Number.

Reply:
Unfortunately this breaks Retire Prior To, as that field would apply to CIDs from different paths, despite declaring that it only works on single path. When we designed the Retire Prior To mechanism, it was important that you can _atomically_ retire all existing CIDs _and_ provide a new CID.

> Endpoints use the sequence number to find the corresponding CID.

Where is this is necessary? You'd use (path ID, CID Sequence number) to find the corresponding CID.

Reply:
That's quite reasonable.

Reply:
Updated and added more definition for the frame fields.

Reply:
I would not change the retire connection ID frame. I think keep a unique Sequence Number for each CID results in something simpler:

* retire CID: the unique CID number is unambiguous.
* retire CID prior: still works.

The main use case for "retire prior" is when CID depend of some key, and the key is renewed. That implies "retire all CID allocated prior to a certain point", and it should be independent of the path attribute of the CID.